packages/go_router/lib/src/redirection.dart

```diff
@@ -10,86 +13,173 @@ import 'matching.dart';
 /// A GoRouter redirector function.
 // TODO(johnpryan): make redirector async
 // See https://github.com/flutter/flutter/issues/105808
-typedef RouteRedirector = RouteMatchList Function(RouteMatchList matches,
-    RouteConfiguration configuration, RouteMatcher matcher,
-    {Object? extra});
+typedef RouteRedirector = Future<RouteMatchList> Function(
+    BuildContext, Future<RouteMatchList>, RouteConfiguration, RouteMatcher,
+    {List<RouteMatchList>? redirectHistory, Object? extra});
 
 /// Processes redirects by returning a new [RouteMatchList] representing the new
 /// location.
-RouteMatchList redirect(RouteMatchList prevMatchList,
-    RouteConfiguration configuration, RouteMatcher matcher,
-    {Object? extra}) {
-  // Store each redirect to detect loops
-  final List<RouteMatchList> redirects = <RouteMatchList>[prevMatchList];
-
-  // Keep looping until redirecting is done
-  while (true) {
-    final RouteMatchList currentMatches = redirects.last;
+Future<RouteMatchList> redirect(
+    BuildContext context,
+    Future<RouteMatchList> prevMatchListFuture,
+    RouteConfiguration configuration,
+    RouteMatcher matcher,
+    {List<RouteMatchList>? redirectHistory,
+    Object? extra}) {
+  return prevMatchListFuture
+      .then<RouteMatchList>((RouteMatchList prevMatchList) {
```

Comment:
Can you use `await` here?

Reply:
Sychronous future can't be used in function that marked async, that is why I refactor the entire redirect pipeline.

Reply:
I don't follow, this works as expected on dartpad:

```dart
import 'package:flutter/foundation.dart';

Future<int> hello() => SynchronousFuture<int>(2);

void main() async {
  print(await hello());
}
```